src/ERC6538Registry.sol

```diff
@@ -1,21 +1,65 @@
-// SPDX-License-Identifier: UNLICENSED
-pragma solidity 0.8.20;
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.23;
 
 import {IERC6538Registry} from "./interfaces/IERC6538Registry.sol";
 
 /// @dev `ERC6538Registry` contract to map accounts to their stealth meta-address. See
 /// [ERC-6538](https://eips.ethereum.org/EIPS/eip-6538) to learn more.
 contract ERC6538Registry is IERC6538Registry {
-  /// @notice Maps a registrant's identifier to the scheme ID to the stealth meta-address.
+  /// @notice Next nonce expected from `user` to use when signing for `registerKeysOnBehalf`.
   /// @dev `registrant` may be a standard 160-bit address or any other identifier.
   /// @dev `schemeId` is an integer identifier for the stealth address scheme.
-  mapping(bytes registrant => mapping(uint256 schemeId => bytes)) public stealthMetaAddressOf;
+  mapping(address registrant => mapping(uint256 schemeId => bytes stealthMetaAddress)) public
+    stealthMetaAddressOf;
+
+  /// @notice A nonce used to ensure a signature can only be used once.
+  /// @dev `user` is the registrant address.
+  /// @dev `nonce` will be incremented after each valid `registerKeysOnBehalf` call.
+  mapping(address user => uint256 nonce) public nonceOf;
+
+  /// @dev EIP-712 Type hash used in `registerKeysOnBehalf`.
+  bytes32 public constant TYPE_HASH =
+    keccak256("EIP712Domain(string name,string version,uint256 chainId,address registryContract)");
+
+  /// @dev Cache the domain separator as an immutable value.
+  bytes32 private immutable _cachedDomainSeparator;
+  /// @dev Store the chain id that `_cachedDomainSeparator` corresponds to, in order to invalidate
+  /// the cached domain separator if the chain id changes.
+  uint256 private immutable _cachedChainId;
```

Comment:
Hmm I'm not sure how I feel about inheriting all this caching and rechecking logic from OZ. It adds a lot of complexity and overhead to the contract vs. simply calculating the domain separator in the constructor, stuffing it and in an immutable variable, and being done with it.

I guess for chainId in particular, the concern is around a chainsplit on one of the networks where these contracts might be deployed, is that right? In which case one of the forks will (presumably) have a new `chainId` afterwards, right?

If that's the case, then why do we go through the trouble of caching—presumably for gas savings in the happy path case? Is all that worth the complexity we're adding to what's supposed to be a really minimal universal contract? I'm not sure.

Other options here would be:

1. Just calculate the domain separator on deploy, save it and use it from an immutable value
2. Just calculate the domain separator on the fly very time using the chainId opcode and `address(this)`

Reply:
Made an attempt to simplify the contract in [b2af4c6](https://github.com/ScopeLift/stealth-address-erc-contracts/pull/5/commits/b2af4c605138ec6f0a8e6361bfa0de28cb140139)

And went with the 1st option. 

```
  constructor() {
    domainSeparator =
      keccak256(abi.encode(TYPE_HASH, "ERC6538Registry", "1.0", block.chainid, address(this)));
  }
```

Reply:
> I guess for chainId in particular, the concern is around a chainsplit on one of the networks where these contracts might be deployed, is that right? In which case one of the forks will (presumably) have a new chainId afterwards, right?

Yea this is the concern. Some of of the opinion that so many things will break in a chain split, that there's no reason to both with the conditional. This is true if you're e.g. a protocol building on lots of deps like USDC, but IMO the conditional is cheap and simple enough that for an independent thing like this registry that we may as well keep it working after a fork

Ref solmate's ERC20 for a clean and minimal reference implementation https://github.com/transmissions11/solmate/blob/0384dbaaa4fcb5715738a9254a7c0a4cb62cf458/src/tokens/ERC20.sol

Reply:
Kept `chainId` out so the contract keeps working even in the event of a chain fork. https://github.com/ScopeLift/stealth-address-erc-contracts/pull/5/commits/a5898bd3ed0e30960e0f17e686d19a073b0ddb62

Reply:
You need to sign over chain ID to prevent replay attacks across chains, so we can't remove that from the signature. The question here is whether to make chain ID have a conditional check [like solmate does](https://github.com/transmissions11/solmate/blob/0384dbaaa4fcb5715738a9254a7c0a4cb62cf458/src/tokens/ERC20.sol#L163), or just keep it fixed based on the chain ID at construction. It's cheap to use the conditional approach, and our ERC contracts would keep working in the event of a fork, so I'd say use the solmate approach

Reply:
Ok, I see chain ID is still in the domain separator so that's good. But the decision on options 1, 2, and 3 still applies

Reply:
Thank you for the detailed explanation! I think I get it now, and the domain seperator will now be re-calculated and the contract will keep working if there's a chain split, like it is in solmate. Updated in : [075b13d](https://github.com/ScopeLift/stealth-address-erc-contracts/pull/5/commits/075b13d1438122e6fe78af48224146f3d41c3cda)